### PR TITLE
Perform signing and verification operations in parallel

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
When we introduced signing at promotion time, the time to promote an image was extended significantly. This PR modifies the image promoter to sign and verify in parallel. This improves the speed significantly. 

Here's a sample promoting two mulitarch images of `bom` from one registry to another:
  
```
BEFORE:
real	8m39.318s
user	4m48.343s
sys	0m36.424s

AFTER:
real	3m9.877s
user	2m10.370s
sys	0m16.066s

```
In order to run, this patch needs release-sdk bumped to include https://github.com/sigstore/cosign/pull/1992 
That PR fixes the TUF client in the cosign libraries to prevent a panic when running in parallel. 

#### Special notes for your reviewer:

/hold
Needs ~https://github.com/sigstore/cosign/pull/1992~ https://github.com/sigstore/sigstore/pull/503 to run, hold for merge and bump of kubernetes-sigs/release-sdk

/cc @cpanato @veronica @saschagrunert @justaugustus 
@kubernetes-sigs/release-engineering 

#### Does this PR introduce a user-facing change?
```release-note
Sign and verification operations at promotion time are now done in parallel
```

### CODE REVIEW

**Comment by puerco** [comment]
Oops, I forgot to update the mocks. PTAL :)

**Comment by puerco** [comment]
Note: e2es will fail until https://github.com/sigstore/sigstore/pull/503 merges and the commit is in our libraries

File: promoter/image/promoter.go [hunk 1]
```diff
@@ -101,6 +102,12 @@ func (p *Promoter) PromoteImages(opts *options.Options) (err error) {
 		return errors.Wrap(err, "activating service accounts")
 	}
 
+	// Prewarm the TUF cache with the targets and keys. This is done
+	// to avoid collisions when signing and verifying in parallel
+	if err := p.impl.PrewarmTUFCache(); err != nil {
+		return errors.Wrap(err, "prewarming TUF cache")
```
Comment (saschagrunert) [inline promoter/image/promoter.go:108]: I'd say for new source code we probably should stick to:
```suggestion
		return fmt.Errorf("prewarming TUF cache: %w", err)
```

File: internal/promoter/image/sign.go [hunk 2]
```diff
@@ -42,46 +45,93 @@ const (
...
-
-		if !isSigned {
-			logrus.Infof("No signatures found for ref %s, not checking", imageRef)
-			continue
-		}
-
-		signedEdges[edge] = edges[edge]
+	signedEdges, err := di.FindSingedEdges(edges)
+	if err != nil {
+		return nil, errors.Errorf("filtering signed edges")
```
Comment (saschagrunert) [inline internal/promoter/image/sign.go:107]:
```suggestion
		return nil, errors.New("filtering signed edges")
```

File: internal/promoter/image/sign.go [hunk 3]
```diff
@@ -42,46 +45,93 @@ const (
...
-			)
+	// Verify the signatures in the edges we are looking at
+	t := throttler.New(maxParallelVerifications, len(signedEdges))
+	for e := range signedEdges {
+		go func(edge reg.PromotionEdge) {
+			var err error
+			logrus.Infof("Verifying signatures of image %s", edge.SrcReference())
+
+			// Check the staged image signatures
+			if _, err = signer.VerifyImage(edge.SrcReference()); err != nil {
```
Comment (saschagrunert) [inline internal/promoter/image/sign.go:118]:
```suggestion
			if _, err := signer.VerifyImage(edge.SrcReference()); err != nil {
```
allows to remove the `var err error` in L114

**Review by saschagrunert — COMMENTED** [review]
Just some nits, otherwise LGTM

File: promoter/image/promoter.go (hunk 1 above)
Comment (puerco) [reply promoter/image/promoter.go:108]: OK, I've moved this while file to native error handling :+1:

File: internal/promoter/image/sign.go (hunk 3 above)
Comment (puerco) [reply internal/promoter/image/sign.go:118]: Thanks, nice catch!

**Pull request merged**